protocols/relay/src/v2/client/handler.rs

```diff
@@ -0,0 +1,628 @@
+        send_back: oneshot::Sender<Result<super::RelayedConnection, ()>>,
+    },
+}
+
+impl fmt::Debug for In {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            In::Reserve { to_listener: _ } => f.debug_struct("In::Reserve").finish(),
+            In::EstablishCircuit {
+                dst_peer_id,
+                send_back: _,
+            } => f
+                .debug_struct("In::EstablishCircuit")
+                .field("dst_peer_id", dst_peer_id)
+                .finish(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum Event {
+    ReservationReqAccepted {
+        /// Indicates whether the request replaces an existing reservation.
+        renewal: bool,
+    },
+    ReservationReqFailed {
+        /// Indicates whether the request replaces an existing reservation.
+        renewal: bool,
+    },
+    OutboundCircuitReqFailed {},
+    /// An inbound circuit request has been denied.
+    InboundCircuitReqDenied {
+        src_peer_id: PeerId,
+    },
+    /// Denying an inbound circuit request failed.
+    InboundCircuitReqDenyFailed {
+        src_peer_id: PeerId,
+        error: std::io::Error,
+    },
+}
+
+pub struct Prototype {
+    local_peer_id: PeerId,
+}
+
+impl Prototype {
+    pub(crate) fn new(local_peer_id: PeerId) -> Self {
+        Self { local_peer_id }
+    }
+}
+
+impl IntoProtocolsHandler for Prototype {
+    type Handler = Handler;
+
+    fn into_handler(self, remote_peer_id: &PeerId, endpoint: &ConnectedPoint) -> Self::Handler {
+        Handler {
+            remote_peer_id: *remote_peer_id,
+            remote_addr: endpoint.get_remote_address().clone(),
+            local_peer_id: self.local_peer_id,
+            queued_events: Default::default(),
+            pending_error: Default::default(),
+            reservation: None,
+            alive_lend_out_substreams: Default::default(),
+            circuit_deny_futs: Default::default(),
+            send_error_futs: Default::default(),
+            keep_alive: KeepAlive::Yes,
+        }
+    }
+
+    fn inbound_protocol(&self) -> <Self::Handler as ProtocolsHandler>::InboundProtocol {
+        inbound_stop::Upgrade {}
+    }
+}
+
+pub struct Handler {
+    local_peer_id: PeerId,
+    remote_peer_id: PeerId,
+    remote_addr: Multiaddr,
+    pending_error: Option<
+        ProtocolsHandlerUpgrErr<
+            EitherError<inbound_stop::UpgradeError, outbound_hop::UpgradeError>,
+        >,
+    >,
+    /// Until when to keep the connection alive.
+    keep_alive: KeepAlive,
+
+    /// Queue of events to return when polled.
+    queued_events: VecDeque<
+        ProtocolsHandlerEvent<
+            <Self as ProtocolsHandler>::OutboundProtocol,
+            <Self as ProtocolsHandler>::OutboundOpenInfo,
+            <Self as ProtocolsHandler>::OutEvent,
+            <Self as ProtocolsHandler>::Error,
+        >,
+    >,
+
+    reservation: Option<Reservation>,
+
+    /// Tracks substreams lend out to the transport.
```

Comment:
Is this meant to be past-tense perhaps?

```suggestion
    /// Tracks substreams lent out to the transport.
```

Reply:
:+1: done in 49a63273889b3670957bedcb2cf9562abcb5057d.